cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,857 @@
+  \tparam CellPerTileDim Cell number inside each tile per dimension
+  \tparam Hash Hash type (lexicographical or morton)
+  \tparam Key Type of the tile/cell hash key
+  \tparam Value Type of the tile/cell No.
+ */
+template <typename MemorySpace, unsigned long long CellPerTileDim = 4,
+          HashTypes Hash = HashTypes::Naive, typename Key = uint64_t,
+          typename Value = uint64_t>
+class SparseIndexSpace
+{
+  public:
+    //! Number of dimensions, 3 = ijk
+    static constexpr int rank = 3;
+    //! Number of bits (per dimension) needed to index the cells inside a tile
+    static constexpr unsigned long long cell_bits_per_tile_dim =
+        bitCount( CellPerTileDim );
+    //! Number of cells inside each tile (per dimension), tile size reset to
+    //! power of 2
+    static constexpr unsigned long long cell_num_per_tile_dim =
+        1 << cell_bits_per_tile_dim;
+    //! Cell mask (per dimension), indicating which part of the tile
+    //! address(binary bits) will index the cells inside each tile
+    static constexpr unsigned long long cell_mask_per_tile_dim =
+        ( 1 << cell_bits_per_tile_dim ) - 1;
+    //! Number of bits (total) needed to index the cells inside a tile
+    static constexpr unsigned long long cell_bits_per_tile =
+        cell_bits_per_tile_dim + cell_bits_per_tile_dim +
+        cell_bits_per_tile_dim;
+    //! Number of cells (total) inside each tile
+    static constexpr unsigned long long cell_num_per_tile =
+        cell_num_per_tile_dim * cell_num_per_tile_dim * cell_num_per_tile_dim;
+    //! Types
+    using key_type = Key;                        // tile hash key
+    using value_type = Value;                    // tile No.
+    static constexpr HashTypes hash_type = Hash; // hash table
+
+    /*!
+      \brief Constructor
+      \param size The size of the block (MPI rank) (Unit: cell)
+      \param capacity Expected capacity of the allocator to store the tiles
+      \param rehash_factor The factor multiplies to the capacity for resizing,
+      when tile nums exceed the capcity
+    */
+    SparseIndexSpace( const std::array<int, rank> size,
+                      const unsigned int capacity, const float rehash_factor )
+        : _block_id_space( size[0] >> cell_bits_per_tile_dim,
+                           size[1] >> cell_bits_per_tile_dim,
+                           size[2] >> cell_bits_per_tile_dim,
+                           1 << bitCount( capacity ), rehash_factor )
+    {
+        std::fill( _min.data(), _min.data() + rank, 0 );
+        std::copy( size.begin(), size.end(), _max.data() );
+    }
+
+    /*!
+      \brief Insert a cell (given a cell ijk, insert the tile where the cell
+      reside in to hash table) \param cell_i cell id in dim-x \param cell_j cell
+      id in dim-y \param cell_k cell id in dim-z
+    */
+    KOKKOS_INLINE_FUNCTION
+    void insertCell( int cell_i, int cell_j, int cell_k ) const
```

Comment:
Yes, I believe it would be useful! Actually, I had this return in the beginning. But later due to my implementation of `insert` I had some data race so I have to remove this return. With the current implementation of `insert` I believe we can add it back,